spec/models/manageiq/providers/vmware/infra_manager/refresher_storage_profile_spec.rb

```diff
@@ -0,0 +1,107 @@
+      let(:storage_location) { '4d3f9f09-38b9b7dc-365d-0010187f00da' }
+
+      it 'handles storage profile deletion' do
+        storage_profile     = StorageProfile.find_by(:ems_ref => storage_profile_ref)
+        profile_assoc_count = StorageProfileStorage.count
+        storage             = Storage.find_by(:ems_ref => 'datastore-953')
+        expect(storage.storage_profiles).to include(storage_profile)
+        expect(storage_profile.storages).to include(storage)
+
+        refresher         = @ems.refresher.new([@ems])
+        target, inventory = refresher.collect_inventory_for_targets(@ems, [@ems])[0]
+
+        inventory[:storage_profile].delete(storage_profile_ref)
+        inventory[:storage_profile_entity].delete(storage_profile_ref)
+
+        hashes = refresher.parse_targeted_inventory(@ems, target, inventory)
+        refresher.save_inventory(@ems, target, hashes)
+
+        storage = Storage.find_by(:location => storage_location)
+        expect(storage.storage_profiles).not_to include(storage_profile)
+        expect(StorageProfileStorage.count).to eql(profile_assoc_count - 1)
+      end
+    end
+
+    context "assocation with VMs/disks" do
+      let(:storage_profile_ref) { '6fe1c7b4-7f7e-4db1-a545-c756e392de62' }
+      let(:storage_profile) { StorageProfile.find_by(:ems_ref => storage_profile_ref) }
+      let(:vm) { Vm.find_by(:ems_ref => 'vm-901') }
+
+      it 'handles storage profile associated vms' do
+        expect(storage_profile.vms_and_templates).to match_array([vm])
+      end
+
+      it 'handles storage profile associated disks' do
+        disk = vm.disks.detect { |d| d.device_type == 'disk' }
+        storage_profile = StorageProfile.find_by(:ems_ref => storage_profile_ref)
+        expect(disk.storage_profile).to eq(storage_profile)
+        expect(storage_profile.disks).to include(disk)
+      end
+
+      it 'clears the association when the storage profile of a VM/Disk is deleted' do
+        disk = vm.disks.detect { |d| d.device_type == 'disk' }  # only 1 disk in this vm
+        expect(vm.storage_profile).to eq(storage_profile)
+        expect(disk.storage_profile).to eq(storage_profile)
+
+        refresher = @ems.refresher.new([@ems])
+        target, inventory = refresher.collect_inventory_for_targets(@ems, [@ems])[0]
+        expect(inventory[:storage_profile_entity].size).to eql(6)
+        inventory[:storage_profile].delete(storage_profile_ref)
+        inventory[:storage_profile_entity].delete(storage_profile_ref)
+
+        hashes = refresher.parse_targeted_inventory(@ems, target, inventory)
+        refresher.save_inventory(@ems, target, hashes)
+        vm.reload
+        disk.reload
+        expect(StorageProfile.find_by(:ems_ref => storage_profile_ref)).to be_nil
+        expect(vm.storage_profile).to be_nil
+        expect(disk.storage_profile).to be_nil
+      end
+
+      it 'clears the association when the storage profile is detached from a VM and its disk' do
```

Comment:
You don't really need to do this, but since you were eager to learn...

I would separate context from example here. Instead of this one example with a ton of setup code and testing all crammed together, I would...

```
context "when the storage profile is detached from a VM and its disk" do
  # <a bunch of setup code, essentially lines 87-96>
  # ...
  before do
    storage_profile.reload
    vm.reload
    disk.reload
  end

  it "clears the association" do
    expect(storage_profile).not_to be_nil
    # etc etc etc
  end
end
```